### PR TITLE
Dev-Worker optimization

### PR DESCRIPTION
**Pull request by jicata**

### CODE REVIEW

File: Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs [hunk 1]
```diff
@@ -20,19 +20,22 @@ public class SubmissionJob : IJob
...
         {
             this.Name = name;
 
             this.logger = LogManager.GetLogger(name);
             this.logger.Info("SubmissionJob initializing...");
 
             this.stopping = false;
-            this.processingSubmissionIds = processingSubmissionIds;
+
+            // this.processingSubmissionIds = processingSubmissionIds;
```
Comment (Innos) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs]: Remove useless comments.

File: Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs [hunk 2]
```diff
@@ -138,7 +147,8 @@ public void Start()
                 }
 
                 // Next line removes the submission from the list. Fixes problem with retesting submissions.
-                this.processingSubmissionIds.Remove(submission.Id);
+                // this.processingSubmissionIds.Remove(submission.Id);
```
Comment (Innos) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs]: Remove useless comments.

File: Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs [hunk 3]
```diff
@@ -46,39 +47,45 @@ public void Start()
...
+
+                            submissions.ForEach(this.submissionsForProcessing.Enqueue);
+                        }
+
+                        retrievedSubmissionSuccessfully = this.submissionsForProcessing.TryDequeue(out submissionId);                     
+                    }
+
+                    if (retrievedSubmissionSuccessfully)
+                    {
+                        this.logger.InfoFormat("Submission №{0} retrieved from database successfully", submissionId);
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs]: `retrieved from database` is not always correct. Maybe you can remove `from database`

File: Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs [hunk 4]
```diff
@@ -46,39 +47,45 @@ public void Start()
...
                 {
-                    submission.Processing = true;
-                    data.SaveChanges();
+                    var retrievedSubmissionSuccessfully = false;
+                    lock (this.submissionsForProcessing)
+                    {
+                        if (this.submissionsForProcessing.IsEmpty)
+                        {
+                            var submissions = data
+                                .Submissions
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs]: ` .Submissions` must be at the line above

File: Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs (hunk 3 above)
Comment (jicata) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs]: Why wouldn't it be always correct? Where else would a submission come from? It's retrieved from the DB and inserted into the `ConcurrentQueue<Submission>` and then extracted from there. I guess a more precise message would be _from queue_ but for the submission to be in the queue it has to be pulled from the database in the first place.

Comment (vladislav-karamfilov) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs]: It could be a file. ;) The word you are looking for is "data store".

Comment (jicata) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs]: Good point :) "data store" is indeed provider independent way to phrase it.

**Pull request merged**